firestore/src/common/settings.cc

```diff
@@ -51,12 +56,59 @@ void Settings::set_host(std::string host) { host_ = firebase::Move(host); }
 
 void Settings::set_ssl_enabled(bool enabled) { ssl_enabled_ = enabled; }
 
+std::shared_ptr<LocalCacheSettings> Settings::local_cache_settings() {
+  if (used_legacy_cache_settings_) {
+    if (is_persistence_enabled()) {
+      local_cache_settings_ = std::make_shared<PersistentCacheSettings>(
+          PersistentCacheSettings::Create().WithSizeBytes(cache_size_bytes()));
+    } else {
+      local_cache_settings_ =
+          std::make_shared<MemoryCacheSettings>(MemoryCacheSettings::Create());
+    }
+  } else if (local_cache_settings_ == nullptr) {
+    local_cache_settings_ = std::make_shared<PersistentCacheSettings>(
+        PersistentCacheSettings::Create());
+  }
+
+  return local_cache_settings_;
+}
+
+void Settings::set_local_cache_settings(const LocalCacheSettings& cache) {
+  if (used_legacy_cache_settings_) {
+    SimpleThrowIllegalState(
+        "Cannot mix set_local_cache_settings() with legacy cache api like "
+        "set_persistence_enabled() or set_cache_size_bytes()");
+  }
+
+  if (cache.kind() == api::LocalCacheSettings::Kind::kPersistent) {
+    local_cache_settings_ = std::make_shared<PersistentCacheSettings>(
+        static_cast<const PersistentCacheSettings&>(cache));
```

Comment:
Yuck to the static_cast. Is there not a better way to do this? They have a common base class so can't local_cache_settings_ be a pointer to the base class type?

Reply:
local_cache_settings_ IS a pointer to the base class. I am not sure how we can avoid this downcast given `cache` is a reference to the base type. Even if I create a copy-ctor for `LocalCacheSettings` in that ctor I'd still need to know which subclass of `LocalCacheSettings` it is and invoke the corresponding copy-ctor for that concrete type?

Reply:
Fair enough. Thanks for considering the feedback. I still hate the static_cast but I can live with it.